### PR TITLE
[Intel MKL] Enabling MKL Conv2D FWD in eager mode

### PR DESCRIPTION
**Pull request by mahmoud-abuzaina**

### CODE REVIEW

**Comment by gbaned** [comment]
@mahmoud-abuzaina Can you please resolve conflicts? Thanks!

**Comment by gzmkl** [comment]
Mahoud is on long vacation and I am covering him for this PR.

Merge conflicts have been addressed.  Thanks!    -GZ

**Comment by gbaned** [comment]
@gzmkl  Still, conflicts are appearing. Can you please resolve? Thanks!

**Comment by gzmkl** [comment]
@gbaned It seems that I resolved but not committed. Now conflicts should be gone. Thank you!

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (jaingaurav) [inline tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Should this be an unordered map instead?

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (jaingaurav) [inline tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: I think we should keep device_name as `string` and call c_str() in the call to SetDeviceName. Otherwise there is no reference to the actual string object.

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (jaingaurav) [inline tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: I believe T is an invalid variable name as per our c++ style guide. Perhaps `data_type`?

File: tensorflow/core/util/mkl_util.h
Comment (jaingaurav) [inline tensorflow/core/util/mkl_util.h]: Please avoid default arguments, it's better to have callers specify them. In fact, I feel like we should just create 2 separate function if there isn't much they are sharing. Maybe same thing with GetTfShape?

Comment (gzmkl) [reply tensorflow/core/util/mkl_util.h]: I will take the suggestion and do refactoring.

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Technically yes.
However, this coding style (using "vector" is consistent with that of  core/graph/mkl_layout_pass.cc which involves most MKL op rewritting. 
Because the size of the vector is only 3; so not much performance impact. 

Thanks!

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: I do not such local variable naming either. So I am going to rename it.

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Will take your suggestion, thanks!

Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: done

File: tensorflow/core/util/mkl_util.h
Comment (gzmkl) [reply tensorflow/core/util/mkl_util.h]: done

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: done

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (penpornk) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Don't you plan on adding more ops/rewrite rules eventually? I think `unordered_map` is better, and we don't need to keep this consistent with the rewrite info (`rinfo_`) in `mkl_layout_pass.cc`. As a matter of fact, I also think that `rinfo_` should be `unordered_map`.

Also, per our style guide, class member names should end with "_" (`mkl_eager_ops_`).

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (penpornk) [inline tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: (This comment will be irrelevant if we change to `unordered_map`.)
We don't really need the variable `result`. We can return true here and return false after the for loop.

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (penpornk) [inline tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]:
```suggestion
  return padding != "EXPLICIT";
```

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (penpornk) [inline tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: I think range-based for loop will make this easier to read:
```c++
for (const auto& attr : attr_list) {
  (*new_mkl_op)->MutableAttrs()->Set(attr.first /* name */, attr.second /* value */);
}
```

File: tensorflow/core/graph/mkl_graph_util.h
Comment (penpornk) [inline tensorflow/core/graph/mkl_graph_util.h]: Can we add `eager_mode` (or `is_eager`) as an op attribute instead? This way we don't need to rename the op just to pass `eager_mode` through template parameter.

**Review by penpornk — CHANGES_REQUESTED** [review]
Sorry for my delay!

**Comment by gzmkl** [comment]
@penpornk  Thanks for your review comments. I will address them today.   -- GZ

File: tensorflow/core/graph/mkl_graph_util.h
Comment (gzmkl) [reply tensorflow/core/graph/mkl_graph_util.h]: @penpornk I have put a comment right before the declaration of "_MklEager", for refactoring purpose in near future (original developer is on a long vacation). I have included your suggestion in the comment.

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Good suggestion and I am testing the suggested change.

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Will make code change

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Will make code change to get rid of the unnecessary variable "result"

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Done with following
  1. renamed mkl_eager_ops by appending "_"
  2. Add a TODO comment in source file for future refactoring per suggestion

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Done with code change

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Done with code change

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Done with code change

File: tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/mkl_eager_op_rewrite.cc]: Done with code change

**Review by penpornk — COMMENTED** [review]
Thank you for the fixes and for adding TO-DOs in the comment! Does this PR need to be in 2.0? If it's urgent, I'm okay with refactoring later. If not, we should at least do the `unordered_map` change in this PR. Making`eager_mode` an op attribute shouldn't need too many modifications either.

File: tensorflow/core/graph/mkl_graph_util.h
Comment (gzmkl) [reply tensorflow/core/graph/mkl_graph_util.h]: I guess that the developer (Mahmoud) wanted to rename the op in order to make sure that the renaming only occurs in pure eager mode  (versus non-eager-mode, all mkl op rewriting is done  thru MKL layout pass (graph/mkl_layout_pass) which is only triggered at the state of
POST_PARTITIONING (which never reach in pure eager mode!).

So, for now, I would recommend that we do refactoring one the original developer is back)
Thanks

Comment (penpornk) [reply tensorflow/core/graph/mkl_graph_util.h]: But why exactly do we need to rename the op, though? Making `eager_mode` an attribute is enough to tell the difference, and is much cleaner (no need to spawn new `_MklEager*` ops every time we add a rewrite rule).

**Review by penpornk — COMMENTED** [review]
For the sake of having MKL Conv2D FWD op in eager mode in TF 2.0, I'm okay with postponing refactorizations to future PRs (if other reviewers are okay) and have no further comments. 

I'll leave the rest to code owners @jaingaurav and @alextp.

**Review by alextp — CHANGES_REQUESTED** [review]
Overall looks good. Can we get benchmark results to show the difference in performance?

File: tensorflow/core/common_runtime/eager/BUILD
Comment (penpornk) [inline tensorflow/core/common_runtime/eager/BUILD]: Nit:
```suggestion
    }) + if_mkl([":mkl_eager_op_rewrite"]),
```

File: tensorflow/core/common_runtime/eager/BUILD
Comment (penpornk) [inline tensorflow/core/common_runtime/eager/BUILD]: Nit: Since the dependency to this target is already guarded by `if_mkl`, I don't think you need `if_mkl` for `srcs` and `copts` here. TensorFlow won't compile this target if it is not in the dependencies.

**Review by penpornk — CHANGES_REQUESTED** [review]
Friendly ping since your goal is to get this and two other eager PRs (https://github.com/tensorflow/tensorflow/pull/30402 and https://github.com/tensorflow/tensorflow/pull/31168) depending on this PR merged by Friday. Merging takes some time so let's try to finish this PR as soon as possible.

File: tensorflow/core/common_runtime/eager/BUILD
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/BUILD]: Thanks, I will follow up this today

File: tensorflow/core/common_runtime/eager/BUILD
Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/BUILD]: @penpornk  Hi Penporn, based on your suggestion I made the change. However, I got 
the following build errors

... ng/private-tensorflow/.tf_configure.bazelrc: --define=tf_api_version=2
ERROR: /ec/fm/disks/nrv_algo_home01/gzhuang/Working/private-tensorflow/tensorflow/core/common_runtime/eager/BUILD:276:1: //tensorflow/core/common_runtime/eager:mkl_eager_op_rewrite: expected value of type 'list(label)' for attribute 'srcs' in 'cc_library' rule, but got "mkl_eager_op_rewrite.cc" (string)
ERROR: /ec/fm/disks/nrv_algo_home01/gzhuang/Working/private-tensorflow/tensorflow/core/common_runtime/eager/BUILD:276:1: //tensorflow/core/common_runtime/eager:mkl_eager_op_rewrite: expected value of type 'list(string)' for attribute 'copts' in 'cc_library' rule, but got "-DINTEL_MKL=1" (string)
ERROR: /ec/fm/disks/nrv_algo_home01/gzhuang/Working/private-tensorflow/tensorflow/core/distributed_runtime/eager/BUILD:59:1: Target '//tensorflow/core/common_runtime/eager:context' contains an error and its package is in error and referenced by '//tensorflow/core/distributed_runtime/eager:eager_service_impl'
ERROR: /ec/fm/disks/nrv_algo_home01/gzhuang/Working/private-tensorflow/tensorflow/core/distributed_runtime/eager/BUILD:59:1: Target '//tensorflow/core/common_runtime/eager:eager_operation' contains an error and its package is in error and referenced by '//tensorflow/core/distributed_runtime/eager:eager_service_impl'
ERROR: /ec/fm/disks/nrv_algo_home01/gzhuang/Working/private-tensorflow/tensorflow/core/distributed_runtime/eager/BUILD:59:1: Target '//tensorflow/core/common_runtime/eager:execute' contains an error and its package is in error and referenced by '//tensorflow/core/distributed_runtime/eager:eager_service_impl'
ERROR: /ec/fm/disks/nrv_algo_home01/gzhuang/Working/private-tensorflow/tensorflow/core/distributed_runtime/eager/BUILD:59:1: Target '//tensorflow/core/common_runtime/eager:tensor_handle' contains an error and its package is in error and referenced by '//tensorflow/core/distributed_runtime/eager:eager_service_impl'
ERROR: Analysis of target '//tensorflow/tools/pip_package:build_pip_package' failed; build aborted: Cannot compute config conditions
INFO: Elapsed time: 6.973s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (218 packages loaded, 6383 targets configured)

Comment (penpornk) [reply tensorflow/core/common_runtime/eager/BUILD]: It sounds like you removed the brackets. Please keep the brackets (`srcs` and `copts` take a list, not a string).
```
srcs = ["mkl_eager_op_rewrite.cc"],
copts = ["-DINTEL_MKL=1"],
```

Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/BUILD]: @penpornk  Thanks, I will try that.

Comment (penpornk) [reply tensorflow/core/common_runtime/eager/BUILD]: If it doesn't work, please feel free to leave it unchanged. This is minor. Please focus on addressing @alextp's comments (benchmarking results).

Comment (gzmkl) [reply tensorflow/core/common_runtime/eager/BUILD]: Give me 15  minutes and I will check in the changes after final round build validation.

File: tensorflow/core/graph/mkl_graph_util.h
Comment (gzmkl) [reply tensorflow/core/graph/mkl_graph_util.h]: done with change

**Review by penpornk — COMMENTED** [review]
LGTM from TF-MKL side. Please continue working with eager code owners to get this PR approved.

**Comment by penpornk** [comment]
I've manually verified that both users contributing to this PR have signed the CLA with their associated emails (not sure why the bot is complaining). Setting CLA to yes.

**Comment by gzmkl** [comment]
@alextp: some benchmark result

Combined with PR #30401 (conv fwd) and #30402 (conv bwd), for resnet50 training performance
is improved from 0.7 images/second to 7 images/second.

**Comment by gzmkl** [comment]
@[eager code owner]: please let me know if you need more changes or more info. Thanks.  GZ

**Comment by gzmkl** [comment]
@penpornk @alextp @mahmoud-abuzaina  
Please approve this PR if you are fine. We have two more related PRs (eager mode conv bwd; 
eager mode matmul) which need to be included in 2.0 release.
Thanks!

**Review by alextp — DISMISSED** [review]

**Comment by penpornk** [comment]
@gzmkl Please resolve conflicts.

**Comment by gzmkl** [comment]
Yes, I am addressing the conflict issue. Problem is resulted from a previous PR (MKL DNN 1.0) integration and I need some time due to API changes. Hopefully I will get it resolved today.

**Comment by gzmkl** [comment]
Conflicts within mkl_conv_ops.cc have been resolved.

**Review by penpornk — DISMISSED** [review]

**Comment by gbaned** [comment]
@gzmkl, @mahmoud-abuzaina  Could you please sign cla ? Thank you.

**Review by penpornk — APPROVED** [review]
@gbaned The email addresses associated with their commits have CLAs (<mahmoud.abuzaina@intel.com> and <guozhong.zhuang@intel.com>). I'm setting CLA to yes.

**Comment by gzmkl** [comment]
@penpornk  Thank you for the approval. It took quite a while to address the merge conflicts.

**Comment by penpornk** [comment]
There are three test failures. `Windows Bazel` and `Windows Bazel GPU` are already-failing tests. 

`Linux GPU` has a [memory corruption in CAPI.Executor_MatMul_CPUAsync](https://source.cloud.google.com/results/invocations/f927c5de-725e-4d13-9e5b-f9a410205c1a/targets/%2F%2Ftensorflow%2Fc%2Feager:c_api_experimental_test_gpu/log) in //tensorflow/c/eager:c_api_experimental_test_gpu.

@alextp @jaingaurav, do you know if this failure is related to the PR?

@gzmkl I'll be unavailable for a few hours so please coordinate with others on this PR in the meanwhile.

**Comment by gzmkl** [comment]
@penpornk  No, this PR only touches Conv2D forward and is not related to "MatMul" or "gpu".

**Review by penpornk — APPROVED** [review]
Let's try pulling it in then.

**Pull request merged**
